From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 22:57:05 -0700
Subject: [PATCH] vbscript: assignment to a property of a built-in object
 (Err.Number, Err.Description)

The built-in objects (Err, Dictionary-like helpers of the language) refused
every DISPATCH_PROPERTYPUT with E_NOTIMPL, so "Err.Number = 5" and
"Err.Description = ..." stopped the script although the setters existed.
The put now reaches them. Err.Number and Err.HelpContext get their setters:
a number up to 65535 is stored as a VBScript error code as Err.Raise does,
other values are stored as they are.

diff --git a/dlls/vbscript/global.c b/dlls/vbscript/global.c
index 9ddb8b7..2a40e16 100644
--- a/dlls/vbscript/global.c
+++ b/dlls/vbscript/global.c
@@ -172,6 +172,7 @@ static HRESULT WINAPI Builtin_Invoke(IDispatch *iface, DISPID id, REFIID riid, L
     const builtin_prop_t *prop;
     VARIANT args_buf[8], *args;
     unsigned argn, i;
+    BOOL put = FALSE;
     HRESULT hres;
 
     TRACE("(%p)->(%ld %s %ld %d %p %p %p %p)\n", This, id, debugstr_guid(riid), lcid, flags, dp, res, ei, err);
@@ -233,16 +234,22 @@ static HRESULT WINAPI Builtin_Invoke(IDispatch *iface, DISPID id, REFIID riid, L
             return E_FAIL;
         }
 
-        FIXME("call put\n");
+        /* the value is the single argument of the setter */
+        if(dp->cArgs != 1 || dp->cNamedArgs != 1 || dp->rgdispidNamedArgs[0] != DISPID_PROPERTYPUT)
+            return MAKE_VBSERROR(VBSE_FUNC_ARITY_MISMATCH);
+#ifdef SG_MUTANT_VBS_NOPUT
         return E_NOTIMPL;
+#endif
+        put = TRUE;
+        break;
     default:
         FIXME("unsupported flags %x\n", flags);
         return E_NOTIMPL;
     }
 
-    argn = arg_cnt(dp);
+    argn = put ? dp->cArgs : arg_cnt(dp);
 
-    if(argn < prop->min_args || argn > (prop->max_args ? prop->max_args : prop->min_args)) {
+    if(!put && (argn < prop->min_args || argn > (prop->max_args ? prop->max_args : prop->min_args))) {
         WARN("invalid number of arguments\n");
         return MAKE_VBSERROR(VBSE_FUNC_ARITY_MISMATCH);
     }
@@ -4516,11 +4523,18 @@ static HRESULT Err_Description(BuiltinDisp *This, VARIANT *args, unsigned args_c
 
 static HRESULT Err_HelpContext(BuiltinDisp *This, VARIANT *args, unsigned args_cnt, VARIANT *res)
 {
+    HRESULT hres;
+
     TRACE("\n");
 
     if(args_cnt) {
-        FIXME("setter not implemented\n");
-        return E_NOTIMPL;
+        int val;
+
+        hres = to_int(args, &val);
+        if(FAILED(hres))
+            return hres;
+        This->ctx->ei.dwHelpContext = val;
+        return S_OK;
     }
 
     return return_int(res, This->ctx->ei.dwHelpContext);
@@ -4539,8 +4553,14 @@ static HRESULT Err_Number(BuiltinDisp *This, VARIANT *args, unsigned args_cnt, V
     TRACE("\n");
 
     if(args_cnt) {
-        FIXME("setter not implemented\n");
-        return E_NOTIMPL;
+        int code;
+
+        hres = to_int(args, &code);
+        if(FAILED(hres))
+            return hres;
+        /* the number of a VBScript error is kept with the facility, as Raise does */
+        This->ctx->ei.scode = (!code || (code & ~0xffff)) ? code : MAKE_VBSERROR(code);
+        return S_OK;
     }
 
     hres = This->ctx->ei.scode;
